backend/crud.py

```diff
@@ -0,0 +1,211 @@
+            if attr.bind_to_schema == -1:
+                s = sch
+            else:
+                s = db.execute(select(Schema).where(Schema.id == attr.bind_to_schema)).scalar()
+            if s is None:
+                raise Exception(f'Cannot bind to nonexistent schema with id {attr.bind_to_schema}')
+            bfk = BoundFK(attr_def_id=ad.id, schema_id=s.id)
+            db.add(bfk)
+    db.commit()
+    return sch
+        
+        
+def create_entity(db: Session, schema_id: int, data: dict) -> Entity:
+    sch: Schema = db.execute(select(Schema).where(Schema.id == schema_id)).scalar()
+    if sch is None:
+        raise Exception(f'Schema with id {schema_id} does not exist')
+    e = Entity(schema_id=schema_id)
+    db.add(e)
+    db.flush()
+
+    for field, value in data.items():
+        attr: Attribute = db.execute(select(Attribute).where(Attribute.name == field)).scalar()
+        if attr is None:
+            raise Exception(f'Attribute with name `{field}` does not exist')
+
+        attr_def: AttributeDefinition = db.execute(
+            select(AttributeDefinition)
+            .where(AttributeDefinition.schema_id == schema_id)
+            .where(AttributeDefinition.attribute_id == attr.id)
+        ).scalar()
+        if attr_def is None:
+            raise Exception(f'There is no attribute definition for schema id {schema_id} and attribute id {attr.id}')
+
+        model: Value = val_map.get(AttrType(attr.type))
+        if model is None:
+            raise Exception(f'No value model found for type `{attr.type}`')
+        
+        caster = type_map.get(AttrType(attr.type))
+        if caster is None:
+            raise Exception(f'No caster found for type `{attr.type}`')
+
+        try:
+            if isinstance(value, list):
+                if not attr_def.list:
+                    raise Exception(f'Attribute {attr.name} on schema {sch.name} cannot hold multiple values')
+                values = [caster(i) for i in value]
+            else:
+                values = [caster(value)]
+        except ValueError as e:
+            raise # can this even happen with validation done before?
+        if attr_def.unique and not attr_def.list:  # TODO this might require a lock
+            exists = db.execute(
+                select(model)
+                .where(model.attribute_id == attr.id)
+                .where(Entity.schema_id == schema_id)
+                .where(model.value == values[0])
+                .join(Entity, Entity.id == model.entity_id)
+            ).scalar()
+            if exists:
+                raise Exception('Non-unique value')
+        for val in values:
+            v = model(value=val, entity_id=e.id, attribute_id=attr.id)
+            db.add(v)
+    db.commit()
+    return e
+
+
+def delete_entity(db: Session, entity_id: int) -> Entity:
+    e = db.execute(select(Entity).where(Entity.id == entity_id)).scalar()
+    if e is None:
+        raise Exception('Should we raise exception here?')
```

Comment:
If requested via the API, the client should receive a 404, if the entity does not exist or is already deleted. So, I'd say an exception is a good choice :+1: 

And, yes, a lock makes sense to me.